libbeat/publisher/pipeline/pipeline.go

```diff
@@ -349,23 +369,96 @@ func (p *Pipeline) ConnectWith(cfg beat.ClientConfig) (beat.Client, error) {
 		acker = nilACKer
 	}
 
-	producer := p.queue.Producer(producerCfg)
-	client := &client{
-		pipeline:     p,
-		isOpen:       atomic.MakeBool(true),
-		eventer:      cfg.Events,
-		processors:   processors,
-		producer:     producer,
-		acker:        acker,
-		eventFlags:   eventFlags,
-		canDrop:      canDrop,
-		reportEvents: reportEvents,
-	}
+	client.acker = acker
+	client.producer = p.queue.Producer(producerCfg)
 
 	p.observer.clientConnected()
+
+	if client.closeRef != nil {
+		p.registerSignalPropagation(client)
+	}
+
 	return client, nil
 }
 
+func (p *Pipeline) registerSignalPropagation(c *client) {
+	p.guardStartSigPropagation.Do(func() {
+		p.sigNewClient = make(chan *client, 1)
+		go p.runSignalPropagation()
+	})
+	p.sigNewClient <- c
+}
+
+func (p *Pipeline) runSignalPropagation() {
+	var channels []reflect.SelectCase
+	var clients []*client
+
+	channels = append(channels, reflect.SelectCase{
+		Dir:  reflect.SelectRecv,
+		Chan: reflect.ValueOf(p.sigNewClient),
+	})
+
+	for {
+		chosen, recv, recvOK := reflect.Select(channels)
+		if chosen == 0 {
+			if !recvOK {
+				// sigNewClient was closed
+				return
+			}
+
+			// new client -> register client for signal propagation.
+			client := recv.Interface().(*client)
+			channels = append(channels,
+				reflect.SelectCase{
+					Dir:  reflect.SelectRecv,
+					Chan: reflect.ValueOf(client.closeRef.Done()),
+				},
+				reflect.SelectCase{
+					Dir:  reflect.SelectRecv,
+					Chan: reflect.ValueOf(client.done),
+				},
+			)
+			clients = append(clients, client)
+			continue
+		}
+
+		// find client we received a signal for. If client.done was closed, then
+		// we have to remove the client only. But if closeRef did trigger the signal, then
+		// we have to propagate the async close to the client.
+		// In either case, the client will be removed
+
+		i := (chosen - 1) / 2
+		isSig := (chosen & 1) == 1
+		if isSig {
+			client := clients[i]
+			client.doClose()
+		}
+
+		// remove:
+		last := len(clients) - 1
+		ch1 := i*2 + 1
+		ch2 := ch1 + 1
+		lastCh1 := last*2 + 1
+		lastCh2 := lastCh1 + 1
+
+		clients[i], clients[last] = clients[last], nil
+		channels[ch1], channels[lastCh1] = channels[lastCh1], reflect.SelectCase{}
+		channels[ch2], channels[lastCh2] = channels[lastCh2], reflect.SelectCase{}
```

Comment:
@urso oh seriously cool reordering here technique and optimization on the slice usage, I had to write it down to verify it but it all make sense. :)